### PR TITLE
added hdwallet compatible with provider-engine for managing multiple …

### PR DESCRIPTION
**Pull request by zmitton**
…accounts

### CODE REVIEW

File: lib/hd_wallet.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+var KeyPair = require('./generators/key_pair');
+var Transaction = require('ethereumjs-tx');
+
+var HDWallet = function(hdSeed, initAccounts = 0) {
+  this.hdSeed = Phrase.toHDPrivateKey(hdSeed);
+  this.keypairs = [];
+  var key;
+  for (var i = 0 ; i < initAccounts ; i++) {
+    key = this.hdSeed.derive(["m","44'", "60'", "0'","0", i].join("/"));
+    this.keypairs[i] = KeyPair.fromPrivateKey(key.privateKey.toBuffer())
```
Comment (pelle) [inline lib/hd_wallet.js]: I would change this to something like 

```javascript
this.hdsigners[i] = new HDSigner(this.hdSeed,i)
```

File: lib/hd_wallet.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  }
+}
+
+HDWallet.prototype.addAccounts = function(numAccounts = 1) {
+  var totalNumAccounts = numAccounts + this.keypairs.length
+  for (var i = this.keypairs.length ; i < totalNumAccounts ; i++) {
+    key = this.hdSeed.derive(["m","44'", "60'", "0'","0", i].join("/"));
+    this.keypairs[i] = KeyPair.fromPrivateKey(key.privateKey.toBuffer())
```
Comment (pelle) [inline lib/hd_wallet.js]: ditto from above

File: lib/hd_wallet.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  }
+};
+
+HDWallet.prototype.hasAddress = function(address, callback) {
+  callback(null, this.indexOfKeypair(address) !== -1)
+};
+
+HDWallet.prototype.getAddress = function(i = 0) {
+    if(!this.keypairs[i]) {return new Error("address " + i + " doesnt exist")}
+    return this.keypairs[i].address;
```
Comment (pelle) [inline lib/hd_wallet.js]:
```javascript
return this.hdsigners[i].getAddress()
```

File: lib/hd_wallet.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+HDWallet.prototype.getAccounts = function(callback = function(e,r){return r}) {
+  callback(null, this.getAddresses());
+}
+
+HDWallet.prototype.getAddresses = function() {
+  var addresses = []
+  for (var i = 0 ; i < this.keypairs.length ; i++) {
+    addresses.push(this.keypairs[i].address);
```
Comment (pelle) [inline lib/hd_wallet.js]:
```javascript
addresses.push(this.getAddress(i))
```

File: lib/hd_wallet.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    if (e)
+      callback(e,null)
+    else
+      callback(null, '0x' + signedTx);
+  });
+};
+
+HDWallet.prototype.signRawTx = function(rawTx, senderAddress, callback) {
+  var index = this.indexOfKeypair(senderAddress);
+  if (index === -1) return callback(new Error("address not managed by this wallet"), null)
```
Comment (pelle) [inline lib/hd_wallet.js]:
```javascript
var signer = this.hdsigners[index]
return signer.signRawTx(rawTx,callback);
```
and remove the rest of this method

**Comment by zmitton** [comment]
@pelle, k that cleaned it up a bit, back to you 👍

File: lib/hd_wallet.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  var key;
+  for (var i = 0 ; i < initAccounts ; i++) {
+    this.hdSigners[i] = new HDSigner(this.hdSeed,i);
+  }
+}
+
+HDWallet.prototype.addAccounts = function(numAccounts = 1) {
+  var totalNumAccounts = numAccounts + this.hdSigners.length
+  for (var i = this.hdSigners.length ; i < totalNumAccounts ; i++) {
+    key = this.hdSeed.derive(["m","44'", "60'", "0'","0", i].join("/"));
```
Comment (pelle) [inline lib/hd_wallet.js]: You can remove this line I think

**Pull request merged**